### PR TITLE
Long-Running Recurring Donations Batch Job Should Not Create Duplicate Opportunities

### PR DESCRIPTION
**Pull request by dcaballeroSf**
# Critical Changes

# Changes
 - We introduced a change that prevents the Recurring Donation batch job from running if the same batch job is already running.

# Issues Closed

# Community Ideas Delivered

# New Metadata
- Custom Labels:
    - commonConcurrentBatchError

# Deleted Metadata

### CODE REVIEW

**Comment by dcaballeroSf** [comment]
Tracking: W-7656037

File: src/classes/CRLP_Batch_Base.cls [hunk 1]
```diff
@@ -655,17 +655,14 @@ public abstract class CRLP_Batch_Base {
      * @param className The name of the skipped batch class to be used in the error message
      */
     protected void logDuplicateBatchError(String className) {
-
-        CRLPBatchException ex = new CRLPBatchException();
-        ex.setMessage(String.format(Label.CRLP_BatchSkipped, new List<String>{className}));
-        ERR_Handler.processError(ex, ERR_Handler_API.Context.CRLP);
-
+	    CRLPBatchException ex = new CRLPBatchException();
```
Comment (npsp-reedestockton) [inline src/classes/CRLP_Batch_Base.cls:658]: Please fix indentation.

File: src/classes/UTIL_BatchJobService.cls [hunk 2]
```diff
@@ -62,6 +62,15 @@ public with sharing class UTIL_BatchJobService {
     public enum ApexJobStatus {
         HOLDING, QUEUED, PREPARING, PROCESSING, COMPLETED, FAILED, ABORTED
     }
+    /******
```
Comment (npsp-reedestockton) [inline src/classes/UTIL_BatchJobService.cls:65]: Please add a blank line before apexdoc.

File: src/classes/RD2_OpportunityEvaluation_BATCH.cls [hunk 3]
```diff
@@ -129,11 +135,20 @@ public without sharing class RD2_OpportunityEvaluation_BATCH
...
     public Database.QueryLocator start(Database.BatchableContext context) {
-        if (RD2_EnablementService.isRecurringDonations2Enabled) {
-            return Database.getQueryLocator(getRDBatchQuery());
-        } else {
+        
+        if (!RD2_EnablementService.isRecurringDonations2Enabled) {
             throw new BatchException(System.Label.RD2_ErrorEnhancedRDBatchJobCannotBeRun);
         }
+        String className = RD2_OpportunityEvaluation_BATCH.class.getName();
+        boolean hasRunningJob = (resultOfHasAnotherTestRunning != null ? resultOfHasAnotherTestRunning : UTIL_MasterSchedulableHelper.hasRunningJob(className));
```
Comment (npsp-reedestockton) [inline src/classes/RD2_OpportunityEvaluation_BATCH.cls:143]: Please capitalize Boolean.  Also, I don't believe the outer parentheses are needed.  Are they?

File: src/classes/RD2_OpportunityEvaluation_TEST.cls [hunk 4]
```diff
@@ -328,6 +328,36 @@ public with sharing class RD2_OpportunityEvaluation_TEST {
...
+
+        List<Opportunity> opps = oppGateway.getRecords(rd);
+        System.assertEquals(1, opps.size(), 'An Opp should exist: ' + opps);
+
+        RD2_OpportunityEvaluation_BATCH batch  = new RD2_OpportunityEvaluation_BATCH(nextCloseDate.addDays(1));
+        batch.resultOfHasAnotherTestRunning = true;
+        runBatch(batch);
+
+        System.assertEquals(1, [SELECT id FROM Error__c].size(), 'Error should have been created');
+    }
```
Comment (npsp-reedestockton) [inline src/classes/RD2_OpportunityEvaluation_TEST.cls:360]: Please skip a line before the following apexdoc.

File: src/classes/RD2_OpportunityEvaluation_TEST.cls [hunk 5]
```diff
@@ -358,6 +388,8 @@ public with sharing class RD2_OpportunityEvaluation_TEST {
             'Close Date on the new Opp should match next Close Date');
     }
 
+
```
Comment (npsp-reedestockton) [inline src/classes/RD2_OpportunityEvaluation_TEST.cls:391]: We often leave an extra empty line or two between test methods and helpers, but between test methods we just leave one empty line.

File: src/classes/RD_RecurringDonations_BATCH.cls [hunk 6]
```diff
@@ -71,7 +76,6 @@ Schedulable {
         query += ' where npe03__Open_Ended_Status__c = \'' + RD_Constants.OPEN_ENDED_STATUS_OPEN + '\'';
         RD_RecurringDonations_BATCH.isBatchButton = true;
     }
-
```
Comment (npsp-reedestockton) [inline src/classes/RD_RecurringDonations_BATCH.cls:74]: Please re-insert this empty line.

File: src/classes/CRLP_Batch_Base.cls (hunk 1 above)
Comment (force2b) [reply src/classes/CRLP_Batch_Base.cls:658]: I think I was the one who suggested moving the `logDuplicateBatchError()` method to the `UTIL_BatchJobService` class to avoid the code duplication.

File: src/classes/RD2_OpportunityEvaluation_BATCH.cls [hunk 7]
```diff
@@ -58,6 +58,12 @@ public without sharing class RD2_OpportunityEvaluation_BATCH
         private set;
     }
 
+    /**
+     * @description Variable used to mock result of hasAnotherTestRunning
+     */
+    @TestVisible
+    private boolean resultOfHasAnotherTestRunning = null;
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/RD2_OpportunityEvaluation_BATCH.cls:65]: Nice refactor suggestion!

File: src/classes/RD2_OpportunityEvaluation_BATCH.cls [hunk 8]
```diff
@@ -129,11 +157,20 @@ public without sharing class RD2_OpportunityEvaluation_BATCH
...
     */
     public Database.QueryLocator start(Database.BatchableContext context) {
-        if (RD2_EnablementService.isRecurringDonations2Enabled) {
-            return Database.getQueryLocator(getRDBatchQuery());
-        } else {
+        
+        if (!RD2_EnablementService.isRecurringDonations2Enabled) {
             throw new BatchException(System.Label.RD2_ErrorEnhancedRDBatchJobCannotBeRun);
         }
+        String className = RD2_OpportunityEvaluation_BATCH.class.getName();
```
Comment (RadSFDeveloper) [inline src/classes/RD2_OpportunityEvaluation_BATCH.cls:164]: The `className` can be removed and `batchClassName` used instead.

File: src/classes/RD2_OpportunityEvaluation_BATCH.cls [hunk 9]
```diff
@@ -58,6 +58,34 @@ public without sharing class RD2_OpportunityEvaluation_BATCH
...
+            return batchClassName;
+        }
+        set;
+    }
+
+    /***
+    * @description Checks if another concurrent batch job is in progress
+    */
+    @TestVisible
+    private Boolean anotherBatchIsInProgress {
```
Comment (RadSFDeveloper) [inline src/classes/RD2_OpportunityEvaluation_BATCH.cls:78]: The Boolean vars should start like a question, ie `is` and `has` and to be consistent with the `RD` batch var, let's rename it to `isBatchInProgress`.

File: src/classes/RD_RecurringDonations_BATCH.cls [hunk 10]
```diff
@@ -62,16 +62,43 @@ Schedulable {
...
     * @description Class constructor
     */
+
+        /***
+    * @description Contains batch Apex class name
+    */
+    private String batchClassName {
+        get {
+            if (batchClassName == null) {
+                batchClassName = RD2_OpportunityEvaluation_BATCH.class.getName();
```
Comment (RadSFDeveloper) [inline src/classes/RD_RecurringDonations_BATCH.cls:75]: The batch class name should be `RD_RecurringDonations_BATCH`.

File: src/classes/RD_RecurringDonations_BATCH.cls [hunk 11]
```diff
@@ -89,11 +116,17 @@ Schedulable {
         // execute() method will not properly update the processed/failure
         // counts (see W-2443023).
 
+        
         if (RD2_EnablementService.isRecurringDonations2Enabled) {
             throw new RecurringDonationsBatchException(System.Label.RD_ErrorLegacyBatchJobCannotBeRun);
-        } else {
-            return Database.getQueryLocator(query);
         }
+        String className = RD_RecurringDonations_BATCH.class.getName();
```
Comment (RadSFDeveloper) [inline src/classes/RD_RecurringDonations_BATCH.cls:123]: The `className` can be removed and `batchClassName` used instead.

File: src/classes/RD_RecurringDonations_TEST2.cls [hunk 12]
```diff
@@ -564,6 +564,39 @@ private with sharing class RD_RecurringDonations_TEST2 {
         ID ApexJobId = Database.executeBatch(rdbatch, 10);
         Test.stopTest();
     }
+    /***
+     * @description Test that not opportunity is created when another job is running
+     */
+    @isTest
+    public static void testBatchDidntRunIfAnotherJobIsRunning() {
```
Comment (RadSFDeveloper) [inline src/classes/RD_RecurringDonations_TEST2.cls:571]: Let's change test name to `shouldNotRunTheBatchIfAnotherJobIsRunning()`.

File: src/labels/CustomLabels.labels [hunk 13]
```diff
@@ -9875,6 +9875,14 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
         <shortDescription>Error shown when errors must be removed before something else can occur.</shortDescription>
         <value>Clear errors before proceeding.</value>
     </labels>
+    <labels>
+        <fullName>stgConcurrentBatchError</fullName>
+        <categories>Settings</categories>
+        <language>en_US</language>
+        <protected>false</protected>
+        <shortDescription>stgConcurrentBatchError</shortDescription>
+        <value>There is a running job for {0} batch</value>
```
Comment (RadSFDeveloper) [inline src/labels/CustomLabels.labels:9884]: @sam-knox , can you please verify the generic error message displayed when a batch is tried to be executed but another same batch is running:
- user clicks [Run Batch] on NPSP Settings page
- user starts a batch at Apex Anon Window at Developer Console 
- the batch starts as the regular scheduled jobs.

Currently we check concurrency for the legacy and enhanced RD batches, but this will probably be expanded to other batches if/where required.

There is `CRLP_BatchSkipped`:
> Today&apos;&apos;s scheduled {0} batch job was skipped because a previous {0} job was still running. The next batch job will run as scheduled. No action is required; this error is for information purposes only."

 but it works when the batch is to be scheduled. I am not sure the message is aligned when user clicks on [Run Batch] at NPSP Settings. Thoughts?

File: src/classes/RD2_OpportunityEvaluation_BATCH.cls [hunk 14]
```diff
@@ -129,11 +157,20 @@ public without sharing class RD2_OpportunityEvaluation_BATCH
...
-        } else {
+        
+        if (!RD2_EnablementService.isRecurringDonations2Enabled) {
             throw new BatchException(System.Label.RD2_ErrorEnhancedRDBatchJobCannotBeRun);
         }
+        String className = RD2_OpportunityEvaluation_BATCH.class.getName();
+        String query = getRDBatchQuery();
+        
+        if (anotherBatchIsInProgress) {
+            BatchException ex = new BatchException();
```
Comment (RadSFDeveloper) [inline src/classes/RD2_OpportunityEvaluation_BATCH.cls:168]: The `BatchException` is not required.

File: src/classes/RD2_OpportunityEvaluation_TEST.cls [hunk 15]
```diff
@@ -328,6 +328,37 @@ public with sharing class RD2_OpportunityEvaluation_TEST {
...
+            .build();
+
+        List<Opportunity> opps = oppGateway.getRecords(rd);
+        System.assertEquals(1, opps.size(), 'An Opp should exist: ' + opps);
+
+        RD2_OpportunityEvaluation_BATCH batch  = new RD2_OpportunityEvaluation_BATCH(nextCloseDate.addDays(1));
+        batch.anotherBatchIsInProgress = true;
+        runBatch(batch);
+
+        System.assertEquals(1, [SELECT id FROM Error__c].size(), 'Error should have been created');
```
Comment (RadSFDeveloper) [inline src/classes/RD2_OpportunityEvaluation_TEST.cls:359]: The `errorGateway` can be used to get error instead of writing SOQL directly. The expected error message should be verified. The similar is applicable to the new legacy batch test in `RD_RecurringDonations_TEST` .

```
        List<Error__c> errors = errorGateway.getRecords();
        System.assertEquals(1, errors.size(), 'One error should be created: ' + errors);
        
        String expectedMessage = String.format(
            System.Label.stgConcurrentBatchError,
            new String[]{ RD2_OpportunityEvaluation_BATCH.class.getName() }
        );
        System.assertEquals(expectedMessage, errors[0].Full_Message__c, 'Error message should match');
```

**Comment by RadSFDeveloper** [comment]
@diego-caballero , when the error is viewed at the error log, the stack trace is not available. When I tested this having applied code review comments yesterday, the stack trace was there. It has something to do with the `BatchException` and setting the error message before it is passed into the `processError()` method. Can you please investigate?

Screen Shot 2020-07-29 at 5 06 38 PM

File: src/classes/RD_RecurringDonations_BATCH.cls (hunk 6 above)
Comment (force2b) [reply src/classes/RD_RecurringDonations_BATCH.cls:74]: This blank line is still needed.

File: src/classes/RD_RecurringDonations_BATCH.cls [hunk 16]
```diff
@@ -89,11 +116,16 @@ Schedulable {
         // execute() method will not properly update the processed/failure
         // counts (see W-2443023).
 
+
```
Comment (force2b) [inline src/classes/RD_RecurringDonations_BATCH.cls:119]: extra blank line not needed here

**Comment by force2b** [comment]
@RadSFDeveloper and @diego-caballero I don't think the Stack Trace on this type of logged entry is needed. Plus, it could be misleading anyway given that it's a logged "message" as opposed to a true code exception.

File: src/labels/CustomLabels.labels (hunk 13 above)
Comment (sam-knox) [reply src/labels/CustomLabels.labels:9884]: Yeah, a very good point @RadSFDeveloper. Let me think on that for a bit and come up with language that works in all these scenarios.

File: src/labels/CustomLabels.labels [hunk 17]
```diff
@@ -9875,6 +9875,14 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
         <shortDescription>Error shown when errors must be removed before something else can occur.</shortDescription>
         <value>Clear errors before proceeding.</value>
     </labels>
+    <labels>
+        <fullName>stgConcurrentBatchError</fullName>
+        <categories>Settings</categories>
+        <language>en_US</language>
+        <protected>false</protected>
+        <shortDescription>stgConcurrentBatchError</shortDescription>
```
Comment (melissabarber) [inline src/labels/CustomLabels.labels:9883]:
```suggestion
        <shortDescription>There is a running job for {0} batch</shortDescription>
```

File: src/labels/CustomLabels.labels [hunk 18]
```diff
@@ -9875,6 +9875,14 @@ If preferred, you can disable Gift Entry and instead use older Batch Gift Entry
         <shortDescription>Error shown when errors must be removed before something else can occur.</shortDescription>
         <value>Clear errors before proceeding.</value>
     </labels>
+    <labels>
+        <fullName>stgConcurrentBatchError</fullName>
+        <categories>Settings</categories>
+        <language>en_US</language>
+        <protected>false</protected>
```
Comment (RadSFDeveloper) [inline src/labels/CustomLabels.labels:9882]: @diego-caballero , all custom labels should be protected: true. Can you please make the change.

Comment (force2b) [reply src/labels/CustomLabels.labels:9882]: Thank you for catching that @RadSFDeveloper!

File: src/labels/CustomLabels.labels (hunk 17 above)
Comment (force2b) [inline src/labels/CustomLabels.labels:9883]: @dcaballeroSf For the `shortDescription` on the new labels, can you put a description of the use of the label rather than just repeat the api name of the label? The `shortDescription` is an internal only comment that helps us understand the usage of our labels.

Comment (RadSFDeveloper) [inline src/labels/CustomLabels.labels:9883]: Also, as @melissabarber  indicated, the short description should be a descriptive text for the label. 
@sam-knox ^

Comment (force2b) [reply src/labels/CustomLabels.labels:9883]: @RadSFDeveloper I was looking for @melissabarber's comment earlier, but couldn't find it in Github, so I added one of my own above.

File: src/labels/CustomLabels.labels (hunk 13 above)
Comment (sam-knox) [reply src/labels/CustomLabels.labels:9884]: @RadSFDeveloper - Since this is going to be a generic error message, can we use the "common" identified here for the label name instead of "stg". Full name would be: "commonConcurrentBatchError"

Comment (RadSFDeveloper) [reply src/labels/CustomLabels.labels:9884]: Thanks @sam-knox , is the message `<value>There is a running job for {0} batch</value>` is going to be changed or it is good as is?

Comment (sam-knox) [reply src/labels/CustomLabels.labels:9884]: @RadSFDeveloper - That's going to change. Sorry, Melissa was just showing me how suggesting mode works on GitHub. I'll post the actual changes in my review.

Comment (RadSFDeveloper) [reply src/labels/CustomLabels.labels:9884]: Giving the feedback in the comment is good as well, because (if I recall correctly) when the suggestion is accepted a new commit is created, meaning another build. If the code review changes are done locally, and committed once to the remote repo (regardless how many local commits), then build run will once.

File: src/classes/UTIL_BatchJobService.cls [hunk 19]
```diff
@@ -95,6 +95,54 @@ public with sharing class UTIL_BatchJobService {
...
+    }
+
+    /***
+    * @description Builds concurrent batch error message
+    * @param className Apex batch class name
+    * @return String
+    */
+    public String getConcurrentBatchMessage(String className) {
+        return String.format(
+            System.Label.commonCriticalError,
```
Comment (RadSFDeveloper) [inline src/classes/UTIL_BatchJobService.cls:141]: This should be `commonConcurrentBatchError` label.

File: src/classes/RD2_OpportunityEvaluation_BATCH.cls [hunk 20]
```diff
@@ -129,11 +157,18 @@ public without sharing class RD2_OpportunityEvaluation_BATCH
     * @description Batch Start Method
     */
     public Database.QueryLocator start(Database.BatchableContext context) {
-        if (RD2_EnablementService.isRecurringDonations2Enabled) {
-            return Database.getQueryLocator(getRDBatchQuery());
-        } else {
+        
+        if (!RD2_EnablementService.isRecurringDonations2Enabled) {
```
Comment (RadSFDeveloper) [inline src/classes/RD2_OpportunityEvaluation_BATCH.cls:161]: @rponti-sforg , can you please also try to ensure that RD2 batch throws an error when RD2 is not enabled , ie this check continues to work as expected.

**Comment by RadSFDeveloper** [comment]
@rponti-sforg , the changes are complete, can you please re-deploy the latest code to the scratch org.
